README.md
# 171Group11

Project description: https://docs.google.com/document/d/11T3zmgd2aG-FoEJEe0ZhL_xSjO3Tcf9v4jUoucUhsyg/edit?usp=sharing

Data: Tesla Stock https://www.kaggle.com/varpit94/tesla-stock-data-updated-till-28jun2021


## How to Run Demo
### Requirements
This demo can only be run locally as we are requesting data from our flask app which is not supported with github pages. <br/>
First, download the project zip onto your computer. <br/>
Then, you will need to have python3, pip, and Flask installed.

For installing pip: https://phoenixnap.com/kb/install-pip-mac <br/>
For installing python3: https://realpython.com/installing-python/ <br/>
For installing flask: You can run
```
pip install Flask
```
in your terminal once you installed pip.

### Running it
After downlaoding the requirements, locate to "project" folder on your terminal. 
Type in:
```
python3 app.py
```
It should say that the debugger is active. On your web browser, you can then type in and navigate to 
```
localhost:3000
```
to use our demo!


Changelog.md
Changelog

**12/1**
Tsung Chieh Chen: Added poly-reg model into pickled as 'model.pkl'

**11/28-11/30**
Alvin Agana: finished demo of polynomial regression model for using date to predict close

**11/24**  
Caitlin Brown: consolidate all linear and poly models into one notebook for ease of interpretation

**11/23**  
McKenzie Hochberg: updated hyperparameter turning to LSTM.ipynb to show results of tuned hyperparameters  
Alvin Agana: added Python Flask app for demo and added form input. Included how to run app in readme  
Caitlin Brown: add Project-Report folder, contains an outline for typesetting our report in LaTeX. Just needs to be filled in the body text later on.

**11/21**  
McKenzie Hochberg: added hyperparameter tuning to LSTM.ipynb

**11/14**   
Dingxian Wang: add upport_Vector_Regression.ipynb skeleton file, to be finished in a few days

**11/13**  
Caitlin Brown: update LSTM.ipynb, working better. Want to try subdividing train/test data by before, after stock starts skyrocketing in recent years.

**11/11 + 11/12**  
Alvin Agana: Created HTML/CSS/JS files, started working on basic outline + scripts + intro content

**11/10**  
Dingxian Wang: add Lin_Poly_Regression.ipynb

**11/9**  
Caitlin Brown: add LSTM.ipynb, it's not working properly yet and needs some debugging

**10/31**  
Caitlin Brown: tidy up Lin_Poly_Regression.ipynb

**10/27**  
Caitlin Brown: add rough draft of Lin_Poly_Regression.ipynb, linear and polynomial regressions

**10/17**  
Dingxian Wang: initialize repository, add README.md, Draft.md


project/old/script.js
(function() {

    'use strict';
    console.log("reading js");

    const navItems = document.querySelectorAll('nav ul:first-child li');
    const navLinks = document.querySelectorAll('nav ul:first-child li a');
    
    navItems.forEach(function(item, i) {
        item.addEventListener("click", function(e) {
            e.preventDefault();
            let targetID = navLinks[i].getAttribute('href');
            let targetAnchor = document.getElementById(targetID);
            let originalTop = Math.floor(targetAnchor.getBoundingClientRect().top) - 70;
            window.scrollBy({ top: originalTop, left: 0, behavior: 'smooth' });
        });
    });

    const backToTopButton = document.querySelector('#group-info');
    backToTopButton.addEventListener("click", function() {
        let targetAnchor = document.getElementById("#intro");
        let originalTop = Math.floor(targetAnchor.getBoundingClientRect().top) - 70;
        window.scrollBy({ top: originalTop, left: 0, behavior: 'smooth' });
    })

    const githubLink = document.querySelector('nav ul:last-child li');
    githubLink.addEventListener("click", function() {
        window.open("https://github.com/Diean233/171Group11");
    });
})();



project/pickled/model.py
import pickle
import numpy as np
import pandas as pd
import financialanalysis as fa
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import PolynomialFeatures
from sklearn.preprocessing import normalize
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.pipeline import make_pipeline

# reading data in
df = pd.read_csv('../../Data/TSLA.csv')
df.Date = pd.to_datetime(df.Date)
df.Date = fa.datetimeToFloatyear(df.Date)

# Using polynomial model from Lin_Poly_Regression-2.s5
train, test = train_test_split(df, test_size=0.2, random_state=20)

# Need to be sorted for polynomial regression to work properly
train = train.sort_values(by=['Date'])
test = test.sort_values(by=['Date'])

X_train, X_test = train['Date'], test['Date']
y_train, y_test = train['Close'], test['Close']

X_train_og = X_train
y_train_og = y_train

# Reshape data (sklearn gets mad if I don't)
X_train = StandardScaler().fit_transform(X_train.values.reshape(-1, 1))
y_train = StandardScaler().fit_transform(y_train.values.reshape(-1, 1))
X_test = StandardScaler().fit_transform(X_test.values.reshape(-1, 1))
y_test = StandardScaler().fit_transform(y_test.values.reshape(-1, 1))

# Degree 25
polyreg = make_pipeline(PolynomialFeatures(
    25), StandardScaler(), LinearRegression())
polyreg.fit(X_train, y_train)

pickle.dump(polyreg, open('model.pkl', 'wb'))
model = pickle.load(open('model.pkl', 'rb'))


project/pickled/app.py
import numpy as np
import pandas as pd
import financialanalysis as fa
import matplotlib.pyplot as plt

from flask import Flask, request, render_template
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split

import pickle

app = Flask(__name__)
model = pickle.load(open('model.pkl', 'rb'))


@app.route('/', methods=['GET'])
def hello_world():

    return render_template('index.html')


@app.route('/', methods=['POST'])
def predict():
    def getDate(date):
        months = {
            "01": "January",
            "02": "February",
            "03": "March",
            "04": "April",
            "05": "May",
            "06": "June",
            "07": "July",
            "08": "August",
            "09": "September",
            "10": "October",
            "11": "November",
            "12": "December",
        }
        # 2021-01-01
        year = date[:4]
        month = date[5:7]
        day = date[8:]

        if (int(day) < 10):
            day = day[1:]

        return months[month] + " " + day + ", " + year

    # reading data in
    df = pd.read_csv('../../Data/TSLA.csv')
    df.Date = pd.to_datetime(df.Date)
    df.Date = fa.datetimeToFloatyear(df.Date)

    # Using polynomial model from Lin_Poly_Regression-2.s5
    train, test = train_test_split(df, test_size=0.2, random_state=20)

    # Need to be sorted for polynomial regression to work properly
    train = train.sort_values(by=['Date'])
    test = test.sort_values(by=['Date'])

    X_train, X_test = train['Date'], test['Date']
    y_train, y_test = train['Close'], test['Close']

    X_train_og = X_train
    y_train_og = y_train

    # getting user input date
    # formatting the input date to be the same as the df
    date = [request.form.get('date')]
    dateArr = [np.array(date)]
    dateArr = pd.to_datetime(dateArr[0])
    dateArr = np.array(fa.datetimeToFloatyear(dateArr[0])).reshape(-1, 1)

    sc = StandardScaler()
    sc.fit(X_train_og.values.reshape(-1, 1))
    dateArr = sc.transform(dateArr)

    y_pred = model.predict(dateArr)
    sc.fit(y_train_og.values.reshape(-1, 1))
    prediction = sc.inverse_transform(y_pred)
    prediction_price = round(prediction[0][0], 2)

    return render_template('index.html', predictedDate=getDate(date[0]), predictedPrice=prediction_price)


if __name__ == '__main__':
    app.run(port=3000, debug=True)
